dll/win32/shell32/shlexec.cpp

```diff
@@ -2986,3 +2986,85 @@ RealShellExecuteW(
                                lphProcess,
                                0);
 }
+
+// The common helper of ShellExec_RunDLLA and ShellExec_RunDLLW
+static VOID
+ShellExec_RunDLL_Helper(
+    _In_opt_ HWND hwnd,
+    _In_opt_ HINSTANCE hInstance,
+    _In_ PCWSTR pszCmdLine,
+    _In_ INT nCmdShow)
+{
+    TRACE("(%p, %p, %s, 0x%X)\n", hwnd, hInstance, wine_dbgstr_w(pszCmdLine), nCmdShow);
+
+    if (!pszCmdLine || !*pszCmdLine)
+        return;
+
+    // '?' enables us to specify the additional mask value
+    ULONG fNewMask = SEE_MASK_NOASYNC;
+    if (*pszCmdLine == L'?') // 1st question
+    {
+        INT MaskValue;
+        if (StrToIntExW(pszCmdLine + 1, STIF_SUPPORT_HEX, &MaskValue))
+            fNewMask |= MaskValue;
+
+        PCWSTR pch2ndQuestion = StrChrW(pszCmdLine + 1, L'?'); // 2nd question
+        if (pch2ndQuestion)
+            pszCmdLine = pch2ndQuestion + 1;
+    }
+
+    WCHAR szPath[2 * MAX_PATH];
+    if (PathProcessCommandAW(pszCmdLine, szPath, _countof(szPath), L'C') == -1)
+        StrCpyNW(szPath, pszCmdLine, _countof(szPath));
+
+    // Split arguments from the path
+    LPWSTR Args = PathGetArgsW(szPath);
+    if (*Args)
+        *(Args - 1) = UNICODE_NULL;
+
+    PathUnquoteSpacesW(szPath);
+
+    // Execute
+    SHELLEXECUTEINFOW execInfo = { sizeof(execInfo) };
+    execInfo.fMask = fNewMask;
+    execInfo.hwnd = hwnd;
+    execInfo.lpFile = szPath;
+    execInfo.lpParameters = Args;
+    execInfo.nShow = nCmdShow;
+    if (!ShellExecuteExW(&execInfo))
+    {
+        DWORD dwError = GetLastError();
+        if (SHELL_InRunDllProcess()) // Is it a RUNDLL process?
+            ExitProcess(dwError); // Terminate it now
+    }
+}
+
+/*************************************************************************
+ *  ShellExec_RunDLLA [SHELL32.358]
+ */
+EXTERN_C
+VOID WINAPI
+ShellExec_RunDLLA(
+    _In_opt_ HWND hwnd,
+    _In_opt_ HINSTANCE hInstance,
+    _In_ PCSTR pszCmdLine,
+    _In_ INT nCmdShow)
+{
+    CStringW strCmdLine = pszCmdLine; // Keep
+    ShellExec_RunDLL_Helper(hwnd, hInstance, strCmdLine, nCmdShow);
+}
+
+/*************************************************************************
+ *  ShellExec_RunDLLW [SHELL32.359]
+ */
+EXTERN_C
+VOID WINAPI
+ShellExec_RunDLLW(
+    _In_opt_ HWND hwnd,
+    _In_opt_ HINSTANCE hInstance,
+    _In_ PCWSTR pszCmdLine,
+    _In_ INT nCmdShow)
+{
+    CStringW strCmdLine = pszCmdLine; // Keep
+    ShellExec_RunDLL_Helper(hwnd, hInstance, strCmdLine, nCmdShow);
```

Comment:
Do you need here to use an intermediate CStringW (that gets casted back to PCWSTR in the call) here?
Isn't it possible to directly pass `pszCmdLine` to `ShellExec_RunDLL_Helper`, since the helper runs with a unicode string in parameter?

Reply:
> Isn't it possible to directly pass pszCmdLine

Yes, it can.